docs/_build/man/goboom.1
.\" Man page generated from reStructuredText.
.
.TH "GOBOOM" "1" "March 28, 2016" "0.2.1" "goboom"
.SH NAME
goboom \- a dmenu wrapper
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
goboom [option]

goobom_run
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
goboom is wrapper around the dmenu.
goboom is the successor and rewrite of xboomx in Go.
.sp
goboom sorts commands to launch according to their launch frequency.
In other words \- if you launch Chromium and KeePassX all the time \- they will appear in the list of commands first.
.SS Build
.sp
goboom is built using Go 1.5 vendoring and godeps.
.SS Install
.sp
goboom is available via AUR as \fIgoboom\-bin\fP\&. For binary downloads see \fI\%https://bintray.com/atriix/generic/goboom/view\fP\&.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
sudo cp goboom /usr/bin
sudo cp goboom_run /usr/bin
mkdir \-p ~/.goboom
cp config.ini.default ~/.goboom/config.ini
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Set your keybinding to \fIgoboom_run\fP\&.
.SH OPTIONS
.INDENT 0.0
.TP
.B \-\-gc
Run garbage collection of the DB
.TP
.B \-\-launcher
Output launcher command
.TP
.B \-\-post
Update ranking DB
.TP
.B \-\-pre
Generate dmenu in
.TP
.B \-\-stats
View DB stats
.UNINDENT
.SH MIGRATION FROM XBOOMX
.sp
goboom will look for a config file in \fI~/.goboom\fP\&.
You will need to will need to convert you old config file to ini\-format and name it \fIconfig.ini\fP, see the bundled default ini\-file for guidence.
.sp
goboom uses a csv\-file instead sqlite as datastore.
You will need to export your xboomx db as so.
At this point there is not a provided tool to do so, but you can easely export it with something like \fI\%sqlitebrowser\fP <\fBhttp://sqlitebrowser.org/\fP>\&.
.sp
The contents of your exported database should look along the lines of this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
name,count
chroimum,13
keepassx,17
gimp,4
.ft P
.fi
.UNINDENT
.UNINDENT
.SH LICENSE
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
goboom \- a dmenu wrapper
Copyright (C) 2016 Victor Häggqvist

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
Victor Häggqvist
.SH COPYRIGHT
2016, Victor Häggqvist
.\" Generated by docutils manpage writer.
.


goboom.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"os/user"
	"path/filepath"
	"sort"
	"strings"

	"github.com/go-ini/ini"
	"github.com/gocarina/gocsv"
	flag "github.com/victorhaggqvist/pflag"
)

var version = "0.0.0" // injected at build time

type Runnable struct {
	Cmd   string `csv:"name"`
	Count int    `csv:"count"`
}

type Config struct {
	DmenuParams string
	Ignore      []string `delim:","`
}

type CmdList []*Runnable

func (c CmdList) Len() int {
	return len(c)
}

func (c CmdList) Swap(i, j int) {
	c[i], c[j] = c[j], c[i]
}

func (c CmdList) Less(i, j int) bool {
	return c[i].Count > c[j].Count
}

var (
	prePhase   bool
	postPhase  bool
	launcher   bool
	stats      bool
	gc         bool
	dbFilePath string
	config     Config
)

func isInIgnoreList(item string) bool {
	for _, i := range config.Ignore {
		if i == item {
			return true
		}
	}
	return false
}

func addIfNotContains(items []string, item string) []string {
	for _, i := range items {
		if i == item {
			return items
		}
	}
	return append(items, item)
}

func openDB() map[string]int {
	if _, err := os.Stat(dbFilePath); err != nil {
		return make(map[string]int)
	}

	list := CmdList{}
	file, _ := os.Open(dbFilePath)
	if err := gocsv.UnmarshalFile(file, &list); err != nil {
		panic(err)
	}

	itemSet := make(map[string]int)
	for _, item := range list {
		itemSet[item.Cmd] = item.Count
	}
	return itemSet
}

func generatePath() []string {
	sysPath := os.Getenv("PATH")
	paths := strings.Split(sysPath, ":")

	pathItems := sort.StringSlice{}

	for _, p := range paths {
		if _, err := os.Stat(p); err == nil {
			files, _ := ioutil.ReadDir(p)
			for _, f := range files {
				if !isInIgnoreList(f.Name()) {
					pathItems = addIfNotContains(pathItems, f.Name())
				}
			}
		}
	}

	sort.Sort(pathItems)
	return pathItems
}

func rankPath(pathItems []string) CmdList {
	itemSet := openDB()

	rankedItems := CmdList{}
	for _, item := range pathItems {
		count, exists := itemSet[item]
		if !exists {
			count = 0
		}
		rankedItems = append(rankedItems, &Runnable{item, count})
	}

	sort.Sort(rankedItems)
	return rankedItems
}

func loadIni() {
	usr, _ := user.Current()
	configPath := filepath.Join(usr.HomeDir, ".goboom")
	iniFile := filepath.Join(configPath, "config.ini")
	dbFilePath = filepath.Join(configPath, "rankdb.csv")

	if _, err := os.Stat(configPath); err != nil {
		err := os.Mkdir(configPath, os.ModePerm)
		if err != nil {
			fmt.Println("failed to create config dir")
		}
	}

	config = Config{
		DmenuParams: "-b -i -nb black -nf orange -sb black -p \">\"",
		Ignore:      []string{"X", "su"},
	}
	if _, err := os.Stat(iniFile); err != nil {
		newCfg := ini.Empty()
		newCfg.NameMapper = ini.TitleUnderscore

		err := ini.ReflectFrom(newCfg, &config)
		if err != nil {
			panic(err)
		}

		newCfg.SaveTo(iniFile)
	} else {
		if err := ini.MapToWithMapper(&config, ini.TitleUnderscore, iniFile); err != nil {
			panic(err)
		}
	}
}

func updateRank(runnable string) {
	itemSet := openDB()

	count, exists := itemSet[runnable]
	if !exists {
		itemSet[runnable] = 1
	} else {
		itemSet[runnable] = count + 1
	}

	writeDB(itemSet)
}

func writeDB(itemSet map[string]int) {
	items := CmdList{}
	for runnable, count := range itemSet {
		items = append(items, &Runnable{runnable, count})
	}

	file, err := os.OpenFile(dbFilePath, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, 0644)
	if err != nil {
		panic(err)
	}
	if err = gocsv.MarshalFile(&items, file); err != nil {
		panic(err)
	}
}

func displayStats() {
	items := openDB()

	sortedList := CmdList{}
	for runnable, count := range items {
		sortedList = append(sortedList, &Runnable{runnable, count})
	}

	sort.Sort(sortedList)

	fmt.Println("Name\tLaunch count")
	for _, i := range sortedList {
		fmt.Printf("%s\t%d\n", i.Cmd, i.Count)
	}
}

func runGC() int {
	items := openDB()

	cleanItems := make(map[string]int)
	removedCount := 0

	for runnable, count := range items {
		if count > 1 {
			cleanItems[runnable] = count
		} else {
			removedCount = removedCount + 1
		}
	}

	writeDB(cleanItems)
	return removedCount
}

func main() {
	flag.BoolVar(&prePhase, "pre", false, "Generate dmenu in")
	flag.BoolVar(&launcher, "launcher", false, "Output launcher command")
	flag.BoolVar(&postPhase, "post", false, "Update ranking DB")
	flag.BoolVar(&stats, "stats", false, "View DB stats")
	flag.BoolVar(&gc, "gc", false, "Run garbage collection of the DB")
	flag.Version = version
	flag.Usage = func() {
		flag.PrintVersion()
		fmt.Print("\nTo actually use goboom execute goboom_run\n\n")
		fmt.Println("Options: ")
		flag.PrintDefaults()
	}
	flag.Parse()

	loadIni()

	if prePhase {
		pathList := generatePath()
		sortedList := rankPath(pathList)
		for _, item := range sortedList {
			fmt.Println(item.Cmd)
		}
	} else if postPhase {
		scanner := bufio.NewScanner(os.Stdin)
		scanner.Scan()
		input := scanner.Text()
		updateRank(input)
		fmt.Print(input)
	} else if launcher {
		fmt.Print("dmenu " + config.DmenuParams)
	} else if stats {
		displayStats()
	} else if gc {
		fmt.Printf("Removed %d items\n", runGC())
	} else {
		flag.Usage()
	}
}
